File: bounce/src/main/java/is/ricko/bounce/endpoint/BounceController.java
package is.ricko.bounce.endpoint;

import is.ricko.bounce.config.BounceConfig;
import is.ricko.bounce.data.BounceLinkGateway;
import is.ricko.bounce.error.NotFoundException;
import is.ricko.bounce.model.BounceAgent;
import is.ricko.bounce.model.BounceLink;
import is.ricko.bounce.util.BounceUtil;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpMethod;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestMethod;

import javax.servlet.http.HttpServletRequest;
import javax.servlet.http.HttpServletResponse;
import java.io.IOException;

import static is.ricko.bounce.util.BounceUtil.chop;

@Controller
public class BounceController {
  private final String cookieComment;
  private final String cookieDomain;
  private final int cookieMaxAge;
  private final String cookieName;
  private final BounceLinkGateway gateway;

  @Autowired
  public BounceController(final BounceConfig config, final BounceLinkGateway gateway) {
    this.cookieName = config.getCookieName();
    this.cookieDomain = config.getDomain();
    this.cookieComment = config.getCookieComment();
    this.cookieMaxAge = config.getCookieMaxAge();
    this.gateway = gateway;
  }

  @RequestMapping(path = {"/{key:^(?!oauth|admin).*$}", "/"}, method = {RequestMethod.GET, RequestMethod.HEAD})
  public void onGet(
    @PathVariable(value = "key", required = false) final String key,
    final HttpServletRequest request,
    final HttpServletResponse response
  ) throws IOException {
    final BounceLink link = requireLink(key);
    final String ua = request.getHeader("User-Agent");
    final BounceAgent agent = gateway.findAgent(ua);
    final boolean isPeek = HttpMethod.HEAD.matches(request.getMethod());
    final String cookie = BounceUtil.getCookie(request, cookieName);
    final String remoteAddr = request.getRemoteAddr();
    gateway.hitOrPeek(isPeek, link, remoteAddr,
      chop(request.getHeader("Referer"), 255),
      chop(ua, 255),
      chop(cookie, 32),
      agent == null ? null : agent.getVendor(),
      agent == null ? null : agent.getMajor(),
      agent == null ? null : agent.getVer(),
      agent == null ? null : agent.getMobile(),
      agent == null ? null : agent.getBot()
    );
    if (isPeek) {
      response.setStatus(HttpStatus.NO_CONTENT.value());
      response.addHeader("Location", link.getTitle());
      response.addHeader("Title", link.getTitle());
      if (link.getHits() != null) response.addHeader("Hits", link.getHits().toString());
      if (link.getPeeks() != null) response.addHeader("Peeks", link.getPeeks().toString());
    } else {
      response.addCookie(BounceUtil.buildCookie(cookieName, cookie, cookieComment, cookieDomain, cookieMaxAge));
      response.sendRedirect(link.getTo());
    }
  }

  private BounceLink requireLink(final String key) {
    return gateway.findByKey(key == null ? "" : key).orElseThrow(() -> new NotFoundException("No such link"));
  }
}

File: bounce/src/main/kotlin/is/ricko/bounce/BounceController.kt
package `is`.ricko.bounce

import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.HttpMethod
import org.springframework.http.HttpStatus
import org.springframework.web.bind.annotation.PathVariable
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestMethod
import org.springframework.web.bind.annotation.RestController
import java.net.Inet4Address
import java.net.InetAddress
import java.net.UnknownHostException
import java.nio.ByteBuffer
import java.util.*
import javax.servlet.http.Cookie
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@RestController
class BounceController @Autowired constructor(
    private val bounceLinkGateway: BounceLinkGateway,
    private val config: BounceConfig
) {
    @RequestMapping(path = ["/{key:^(?!oauth|admin).*$}", "/"], method = [RequestMethod.GET, RequestMethod.HEAD])
    fun bounceGet(
        @PathVariable(value = "key", required = false) key: String?,
        request: HttpServletRequest,
        response: HttpServletResponse
    ) {
        val link: BounceLink = requireLink(key)
        val ua: String? = request.getHeader("User-Agent")
        val agent: BounceAgent? = bounceLinkGateway.agent(ua)
        val isPeek: Boolean = HttpMethod.HEAD.matches(request.method)
        val cookie: String? = getCookie(request)
        bounceLinkGateway.hitOrPeek(
            isPeek = isPeek,
            link = link,
            ipv4 = ipv4FromAddr(request.remoteAddr),
            ref = chop(request.getHeader("Referer"), 255),
            ua = chop(ua, 255),
            cookie = chop(cookie, 32),
            ua_vendor = agent?.vendor,
            ua_major = agent?.major,
            ua_ver = agent?.ver,
            mobile = agent?.mobile,
            bot = agent?.bot
        )
        if (isPeek) {
            response.status = HttpStatus.NO_CONTENT.value()
            response.addHeader("Location", link.to)
            response.addHeader("Title", link.title)
            if (link.hits != null) response.addHeader("Hits", link.hits.toString())
            if (link.peeks != null) response.addHeader("Peeks", link.peeks.toString())
        } else {
            response.addCookie(buildCookie(cookie))
            response.sendRedirect(link.to)
        }
    }

    private fun requireLink(key: String?): BounceLink {
        return bounceLinkGateway.findByKey(key ?: "") ?: throw NotFoundException("No such link")
    }

    private fun buildCookie(value: String?): Cookie {
        val sessionId: String = if (value == null || value.isBlank()) {
            val uuid: UUID = UUID.randomUUID()
            val bytes: ByteArray = ByteBuffer.wrap(ByteArray(16))
                .putLong(uuid.mostSignificantBits)
                .putLong(uuid.leastSignificantBits)
                .array()
            Base64.getUrlEncoder().encodeToString(bytes)
        } else {
            value
        }
        val outCookie = Cookie(config.cookieName, sessionId)
        outCookie.comment = config.cookieComment
        outCookie.domain = config.domain
        outCookie.maxAge = config.cookieMaxAge
        outCookie.isHttpOnly = true
        return outCookie
    }

    private fun getCookie(request: HttpServletRequest): String? {
        val cookieName = config.cookieName
        val cookies: Array<out Cookie>? = request.cookies
        if (cookies == null || cookies.isEmpty()) return null
        for (cookie in cookies) {
            if (cookieName == cookie.name) {
                return cookie.value
            }
        }
        return null
    }
}

fun ipv4FromAddr(addr: String): Int? {
    return try {
        val inetAddress: InetAddress = Inet4Address.getByName(addr)
        ByteBuffer.wrap(inetAddress.address).int
    } catch (ex: UnknownHostException) {
        null
    }
}

fun chop(s: String?, c: Int): String? {
    return if (s == null || s.isBlank()) null else if (s.length <= c) s else s.substring(0, c)
}
